File: backend/src/sandbox/api/tool/write.py
"""Internal implementation for the public sandbox file-write verb."""

from __future__ import annotations

from audit.base import AuditSink
from sandbox.api.tool._daemon_payload import daemon_request_identity
from sandbox.api.tool._operation_audit import run_audited_operation
from sandbox.api.tool._result_projection import guarded_result_from_daemon_response
from sandbox.api.timeouts import WRITE_FILE_TIMEOUT_S
from sandbox.api.transport import DAEMON_OP_WRITE_FILE, DaemonSandboxTransport, SandboxTransport
from sandbox._shared.models import WriteFileRequest, WriteFileResult


async def write_file(
    sandbox_id: str,
    request: WriteFileRequest,
    *,
    audit_sink: AuditSink | None = None,
    transport: SandboxTransport | None = None,
) -> WriteFileResult:
    """Write one UTF-8 file through sandbox-local OCC."""
    daemon_transport = transport or DaemonSandboxTransport()

    async def _call() -> WriteFileResult:
        payload = daemon_request_identity(request) | {
            "path": request.path,
            "content": request.content,
            "description": request.default_description(f"write {request.path}"),
            "overwrite": request.overwrite,
        }
        response = await daemon_transport.call(
            sandbox_id,
            DAEMON_OP_WRITE_FILE,
            payload,
            timeout=WRITE_FILE_TIMEOUT_S,
        )
        return guarded_result_from_daemon_response(WriteFileResult, response)

    return await run_audited_operation(
        audit_sink=audit_sink,
        sandbox_id=sandbox_id,
        operation="write_file",
        caller=request.caller,
        payload={"path": request.path},
        call=_call,
    )


__all__ = ["write_file"]

File: agent-core/crates/eos-sandbox-api/src/tool_api/write.rs
//! Pure `write_file` helper: build payload → call transport → parse envelope.

use eos_types::SandboxId;
use serde_json::Value;

use crate::error::SandboxApiError;
use crate::models::{WriteFileRequest, WriteFileResult};
use crate::ops::DaemonOp;
use crate::timeouts::WRITE_FILE_TIMEOUT_S;
use crate::tool_api::parse::{daemon_request_identity_fields, parse_write_file_result};
use crate::transport::SandboxTransport;

/// Write one UTF-8 file through sandbox-local OCC.
pub async fn write_file(
    transport: &dyn SandboxTransport,
    sandbox_id: &SandboxId,
    request: &WriteFileRequest,
) -> Result<WriteFileResult, SandboxApiError> {
    let mut payload = daemon_request_identity_fields(&request.base);
    payload.insert("path".to_owned(), Value::String(request.path.clone()));
    payload.insert("content".to_owned(), Value::String(request.content.clone()));
    payload.insert(
        "description".to_owned(),
        Value::String(
            request
                .base
                .description_or(&format!("write {}", request.path)),
        ),
    );
    payload.insert("overwrite".to_owned(), Value::Bool(request.overwrite));
    let response = transport
        .call(
            sandbox_id,
            DaemonOp::WriteFile,
            payload,
            WRITE_FILE_TIMEOUT_S,
        )
        .await?;
    parse_write_file_result(&response)
}
